### DISCUSSION TITLE
Como implementar streamBuilders (conectadas a uma única stream) em múltiplas telas para ter os dados atualizados constantemente em todas as telas ?

### DISCUSSION THREAD

**Discussion by EuHigorBarbosa** [Flutterando/forum#87] [Perguntas e Respostas] [answered]
Pessoal, tudo bem?
Estou precisando de uma ajuda num programa que estou desenvolvendo. Basicamente eu uso uma stream em um singleton e quero mostrar os dados dessa stream em várias telas usando o streamBuilder. 

Mas estou enfrentando o seguinte problema: Imaginem um programa com 3 telas: A, B e C. Na tela C eu atualizo a stream e mostro os dados do snapshot na tela por meio do StreamBuilder.  Na tela B eu também mostro os dados da mesma stream usando o streamBuilder. Porém, depois de alterar os valores na tela C e retornar à tela B acontece que os valores mostrados pelo streamBuilder da tela B não estão atualizados.

E quando eu vou da B para A e volto para B os valores se atualizam. 

Eu quero que os valores fiquem atualizados o tempo todo em todas as telas. 

Eu já tentei implementar o get_it e não funcionou. Com o get_it o resultado é o mesmo. 

Vocês podem ter acesso ao [git que construi pra discutir o problema aqui: ](https://github.com/EuHigorBarbosa/stream_builder_test.git)

Eu também fiz uma [pergunta no stackOverflow: ](https://stackoverflow.com/questions/74577642/how-to-implement-streambuilder-in-multiples-pages-to-have-an-updated-data-in-all) -> lá tem os códigos descritos

E aqui tem um [vídeo do youtube](https://youtu.be/-TI8sa4I650) que fiz para mostrar a questão ...

Obrigado a todos...

**Comment by eduardoflorence** [comment, marked as answer]
Olá @EuHigorBarbosa,

1. No seu git tem um erro em `PageCC`. A linha `SeqDepService seqDepService = SeqDepService();` deveria estar dentro da classe que estende `State` e ela está entre as classes. (isso não resolve o problema)
2. Para o seu cenário não existe necessidade de utilizar `MultiStreamController`, pois um `StreamController.broadcast()` já resolve. Você pode simplificar seu `SeqDepServiceFFmpeg` para ficar assim:

```dart
class SeqDepServiceFFmpeg implements SeqDepService {
  static SeqDepServiceFFmpeg? _instance;
  SeqDepServiceFFmpeg._(); 

  factory SeqDepServiceFFmpeg(){
      _instance ??= SeqDepServiceFFmpeg._();
      return _instance!;
  }

  final List<SeqDepEntity> _activeListSeqDep = [];
  final StreamController<List<SeqDepEntity>> _controller = StreamController<List<SeqDepEntity>>.broadcast();
   
  @override
  List<SeqDepEntity> get activeSeqDepList => _activeListSeqDep;

  @override
  Stream<List<SeqDepEntity>> get activeSeqDepListStream => _controller.stream;
  
  @override
  void addActiveSeqDep( SeqDepEntity seqDep) {
    _activeListSeqDep.add(seqDep);
    _addInStream(_activeListSeqDep);
  }
  
  void _addInStream (List<SeqDepEntity> activeEntities){
    _controller.sink.add(activeEntities);
  }

}
```

3. Nos `StreamBuilder` acrescente também a propriedade `initialData` com o valor `seqDepService.activeSeqDepList`, pois senão quando o stream já tem um valor, ao entrar numa página o StreamBuilder não carrega esse valor e fica esperando um novo valor chegar.

**Reply by EuHigorBarbosa** [reply]
Eduardo Florence...meu muito obrigado por me ajudar. Eu fiz as modificações que você sugeriu e elas funcionaram. 
Sou grato a você por ter me ajudado. Precisando de mim pode me acionar que se eu puder lhe retribuir ficarei muito feliz.

Contudo ainda tenho uma dúvida quanto ao funcionamento das stream quando se implementa streamBuilders em varias telas. 

No caso desse mini programa que criei existem 3 telas. Quando eu retorno da tela C para a B por meio do pop() não há um novo build da tela B. Mas eu achava que as stream eram tão poderosas que elas atualizavam o widget sem a necessidade de rebuildar a tela B. 

Em outras palavras, eu esperava o comportamento de uma mudança instantânea nos valores de todas os streamBuilders que fizessem uso dessa stream e estivessem ativas na árvore de widgets. Esperava esse comportamento sem o uso de initialValue. 

Então minha pergunta é: Qual é o comportamento esperado dos vários streamBuilders (espalhados em várias telas) que estão conectadas a uma mesma stream? (em outras palavras...qual o comportamento das streamBuilders em telas que não estão em foco do usuário?)

**Reply by eduardoflorence** [reply]
Olá @EuHigorBarbosa,
    O comportamento é exatamente esse que você espera, ou seja, haverá a mudança instantânea em todos os StreamBuilder que utilizam o stream. Se você não colocar o `initialData` na tela B, quando você retorna para ela vindo da tela C, o valor tem que estar atualizado. O `initialData` só é útil quando você entra num página nova, sem ela estar na pilha de navegação, pois já mostra o valor atual da Stream, por exemplo, quando você retorna para a página B e depois vai para a página C de novo. Sem o `initialData` na página C, você não conseguiria ver o valor atual do stream e teria que clicar para colocar um novo valor no fluxo.
   Eu fiz um exemplo baseado no seu no DartPad para você acompanhar melhor: https://dartpad.dev/24e4a14f446eb32cb2f2d5b26596a05c

**Reply by EuHigorBarbosa** [reply]
Muito obrigado por me ajudar. Depois que tive essa dúvida eu estudei a fundo o StreamBuilder por meio desse artigo aqui:
https://andrious.medium.com/decode-streambuilder-e60948629d8e

E fiz um exemplo reescrevendo o seu DartPad: (Editado - agora com o link correto)
https://dartpad.dev/1de15101b4161db670c98a5ac2439cbe

Nesse novo exemplo eu confrontei as abordagens para mostrar os dados na UI...utilizando hasData e hasError e abordagem utilizando os valores de ConnectionState.  Fiz isso colocando os 2 tipos de montagem de StreamBulder nas telas BB e CC...para ver em tempo real a diferença. Também montei uma stream autônoma com um timer para vermos as mudanças á medida que o tempo passa.

Vi que essa questão do connectionState não se aplica ao meu caso...Por isso vou mostrar os dados utilizando o hasData no meu projeto principal. 

Muito obrigado pela ajuda...foi de grande valia.
